lib/Parse/ParseExpr.cpp

```diff
@@ -3127,19 +3143,88 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     return status;
 
   // Parse the closure.
-  ParserResult<Expr> closure =
-    parseTrailingClosure(SourceRange(leftLoc, rightLoc));
-  status |= closure;
-  if (closure.isNull())
-    return status;
+  status |=
+      parseTrailingClosures(isExprBasic, SourceRange(leftLoc, rightLoc),
+                            trailingLBrace, trailingRBrace, trailingClosures);
+  return status;
+}
 
-  // Record the trailing closure.
-  trailingClosure = closure.get();
+/// Check whether upcoming trailing closure actually represents a block of
+/// labeled trailing closures e.g.
+/// foo() {
+///  <label-1>: { ... }
+///  ...
+/// }
+static bool isBlockOfMultipleTrailingClosures(bool isExprBasic, Parser &P) {
+  assert(isValidTrailingClosure(isExprBasic, P));
+  Parser::BacktrackingScope backtrack(P);
+  P.consumeToken(tok::l_brace);
 
-  return status;
+  if (!P.Tok.canBeArgumentLabel() || P.Tok.is(tok::kw__))
+    return false;
+
+  auto text = P.Tok.getText();
+  if (text[0] == '$')
+    return false;
+
+  P.consumeToken(); // Consume Label text.
+
+  if (!P.Tok.is(tok::colon))
+    return false;
+
+  P.consumeToken(); // Consume `:`
+
+  // If the next token after the label is not a left brace
+  // that means what follows is not a valid closure, which
+  // in turn means it can't be a labeled block.
+  if (!P.Tok.is(tok::l_brace))
+    return false;
+
+  return isValidTrailingClosure(isExprBasic, P);
 }
 
-ParserResult<Expr> Parser::parseTrailingClosure(SourceRange calleeRange) {
+ParserStatus Parser::parseMultipleTrailingClosures(
+    SourceLoc &LBrace, SourceLoc &RBrace,
+    SmallVectorImpl<TrailingClosure> &closures) {
+  // Consume '{' of the trailing closure
+  LBrace = consumeToken();
+
+  // There could N labeled closures depending on the number of arguments.
+  do {
+    if (!(Tok.canBeArgumentLabel() && peekToken().is(tok::colon)))
+      return makeParserError();
```

Comment:
Could you keep the already-parsed-closures in the AST somehow? Otherwise, many SourceKit functionalities will be messed up because it cannot know the structure of the source code.

Reply:
Also, when you detect errors in the braces, you should skip to `}` or next stmt (using `skipUntilDeclStmtRBrace()`), and hopefully parse `}` as the `RBrace`.

Reply:
This method is used to peek ahead and check whether what follows represents multiple closures, if not it’s just going to parse it as a regular closure, it doesn’t affect the state of parser or drop already parsed nodes...

Reply:
I think such preservation already happens because closures array with be still returned but with error status. I am going to add skipping though

Reply:
For example:
```
doSomething() {
  labelA: { ... }
  1 + 2
}
```
In this case, `isBlockOfMultipleTrailingClosures()` returns true, right?. We want to skip `1 + 2`, and keep other parts in the AST.

Reply:
Sure, it does make sense to skip all of the invalid bits!

Reply:
Fixed!